rockets/jsonrpc/receiver.h

```diff
@@ -120,14 +120,45 @@ class Receiver
     template <typename Params>
     void bind(const std::string& method, std::function<Response(Params)> action)
     {
-        bind(method, [action](const std::string& request) {
+        bind(method, [action](Request request) {
             Params params;
-            if (!from_json(params, request))
+            if (!from_json(params, request.message))
                 return Response::invalidParams();
             return action(std::move(params));
         });
     }
 
+    /**
+     * Bind a method to a response callback with templated request parameters
+     * and templated return type.
+     *
+     * The parameters object must be deserializable by a free function:
+     * from_json(Params& object, const std::string& json). The return type must
+     * be serializable by a free function: std::string to_json(const RetVal&)
+     *
+     * @param method to register.
+     * @param action to perform.
+     * @throw std::invalid_argument if the method name starts with "rpc."
+     */
+    template <typename Params, typename RetVal>
+    void bind(const std::string& method, std::function<RetVal(Params)> action)
+    {
+        bind(method, [action](Request request) {
+            Params params;
+            if (!from_json(params, request.message))
+                return Response::invalidParams();
+            try
+            {
+                auto ret = action(std::move(params));
+                return Response{to_json(ret)};
+            }
+            catch (const std::runtime_error& e)
+            {
+                return Response{e.what(), -42};
```

Comment:
we could also have our exception sub-class which incorporates an error code to give more flexibility to the user. Possibly a signed int to forbid all conflict with json-rpc reserved codes which are negative.